tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -25,25 +29,190 @@ pub struct CudaVec<T: Numeric> {
 }
 
 impl<T: Numeric> CudaVec<T> {
+    pub fn new(len: usize, stream: &CudaStream) -> Self {
+        let vec = unsafe { Self::new_async(len, stream) };
+        stream.synchronize();
+        vec
+    }
     /// # Safety
     ///
-    /// - `ptr` __must__ be a valid device pointer to an array of `len` elements of type `T`
-    pub unsafe fn new(ptr: CudaPtr, len: usize, device: CudaDevice) -> Self {
+    /// - `stream` __must__ be synchronized to guarantee computation has finished
+    pub unsafe fn new_async(len: usize, stream: &CudaStream) -> Self {
+        let size = len as u64 * std::mem::size_of::<T>() as u64;
+        let ptr = CudaPtr {
+            ptr: cuda_malloc_async(size, stream.as_c_ptr()),
+            device: stream.device,
+        };
+        cuda_memset_async(ptr.ptr, 0u64, size, stream.as_c_ptr());
+
         Self {
             ptr,
             len,
-            device,
+            device: stream.device,
             _phantom: PhantomData,
         }
     }
+
     /// # Safety
     ///
     /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
     ///   not be dropped until stream is synchronised
-    pub unsafe fn from_async(h_data: &Vec<T>, stream: &CudaStream) -> Self {
-        let mut d_data = stream.malloc_async(h_data.len() as u32);
-        stream.copy_to_gpu_async(&mut d_data, h_data.as_slice());
-        d_data
+    pub unsafe fn from_cpu_async(src: &[T], stream: &CudaStream) -> Self {
+        let mut res = Self::new_async(src.len(), stream);
+        // We have to check that h_data is not empty, because cuda_memset with size 0 is invalid
+        if !src.is_empty() {
+            res.copy_to_gpu_async(src, stream);
+        }
+        res
+    }
+
+    /// # Safety
+    ///
+    /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
+    ///   not be dropped until stream is synchronised
+    pub unsafe fn memset_async(&mut self, value: T, stream: &CudaStream)
+    where
+        T: Into<u64>,
+    {
+        let size = self.len() * std::mem::size_of::<T>();
+        // We check that self is not empty to avoid invalid pointers
+        if size > 0 {
+            cuda_memset_async(
+                self.as_mut_c_ptr(),
+                value.into(),
+                size as u64,
+                stream.as_c_ptr(),
+            );
+        }
+    }
+
+    /// Copies data from slice into `CudaVec`
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_to_gpu_async(&mut self, src: &[T], stream: &CudaStream)
```

Comment:
This should be named `copy_from_cpu_async` since it `&mut self` that receives data from cpu slice

Reply:
Indeed